### ISSUE TITLE
Why doesn't the device plugin pass the containerID parameter

### ISSUE THREAD

**Issue by tiantain305** [kubernetes/kubernetes#126148]
### What would you like to be added?

Without containerID, I cannot perform secondary preferences for devices in the GetPreferredAllocation interface

### Why is this needed?

Without containerID, I cannot perform secondary preferences for devices in the GetPreferredAllocation interface

**Comment by pranav-pandey0804** [comment]
Hello @tiantain305,
Could you provide more details on how the lack of `containerID` in the device plugin's `GetPreferredAllocation` interface affects your use cases?

**Comment by pranav-pandey0804** [comment]
This seems critical for use cases requiring optimized device selection or handling of numerous small, stateful operations.
what do you think?

**Comment by pranav-pandey0804** [comment]
also, i couldn't find any existing discussions or issues address this, i would like to understand if there are any underlying reasons for not including containerID that we should consider.

**Comment by pranav-pandey0804** [comment]
@tiantain305 We do need to find out if this enhancement aligns with the broader goals for device plugins in Kubernetes.
Understanding whether such a change is feasible and valuable to the community would be the first step.

**Comment by pranav-pandey0804** [comment]
I think **SIG Scheduling** could also have an interest in this issue due to its implications for resource allocation strategies and how these might affect scheduling decisions.

**Comment by ffromani** [comment]
/triage needs-information

@tiantain305 as mentioned in previous comments, could you please elaborate a bit your use case and what you will need `containerID` for?

**Comment by guojianyu** [comment]
@ffromani  @pranav-pandey0804    The existing preference and allocation interface of the device plugin can only accomplish the allocation of resource capacity, if the pod or container cares about the underlying hardware or the existing mechanism of the specified hardware will not be able to do so. Expect kubelet to pass pod or container information to the device plugin so it can do more refined resource allocation.

**Comment by ffromani** [comment]
@guojianyu I think I can see use cases, but still we'll need a bit more exploration, and quite possibly discussion in sig-node, likely a KEP. Adding some example flow would help narrowing down the flows and the changes.

**Comment by tiantain305** [comment]
@ffromani I am writing a device management component using the device plugin, and I want to use GetPreferredAllocation
 interface provides flexible deviceids for kubelet instead of using the deviceids selected by kubelet. For example, if I register four deviceids {0,1,2,3} for kubelet, when I need to apply for two devices, kubelet may default to 0,3 for me. However, I actually feel that 1,2 is more in line with my expectations. The current device plugin does not provide containerID, so I cannot filter deviceID for a specific container. I will provide an example for discussion later

**Comment by ffromani** [comment]
@tiantain305 thanks for elaborating, adding more details helps. Thing is, the kubelet is supposed to own the resource (including device) selection here, and the topology (NUMA...) alignment. This logic is currently part of the kubelet. Moving this logic elsewhere is something there's general interest into, but definetely requires discussion

**Comment by guojianyu** [comment]
@ffromani @tiantain305 
The following scenarios are unavailable:
1: A pod requires both GPU and FPGA resources, and kubelet first calls the GPU device plugin to complete resource allocation. Based FPGA device plugin extension GetPreferredAllocation wants to know the resource information of the GPU and select a better resource according to its hardware topology information.
(requirement: kubelet needs to pass the container's allocated device information to each device plugin)

2: A pod specifies the FPGA0 device on the allocated node, and device information is passed by annotaion of the pod.
 (requirement: kubelet needs to pass pod information to the device plugin)

3: The pod1 cares about the device resource information of the pod2.
 (requirement: kubelet provides an interface for querying device resources on the same node)

Also, Is it appropriate to integrate topology algorithms into kubelet? Can it move down to device plugin? Like scheduler plugin, device plugin can choose whether to enable the default algorithm.

**Comment by tiantain305** [comment]
@ffromani  I agree with you, Changes need to be realistic and discussed carefully. My personal opinion is why can't  pass containerID between kubelet and the device plugin to allow for a freer allocation of resources

**Comment by alculquicondor** [comment]
/remove-sig scheduling

I don't see how the information about containerID has anything to do with scheduling. Please clarify if you wish to re-add `sig/scheduling`.
